### PR TITLE
Support for IaaS-native persistent disk growing

### PR DESCRIPTION
**Pull request by bgandon**
Hi folks,

We're submitting here our work on the BOSH-Agent for it to properly handle IaaS-native persistent disk growing–indeed we've deliberately put disk shrinking out of the scope.

Let's discuss about merging this!

#### Notice: this PR depends on 
- cloudfoundry/bosh-utils#80
- Some refactoring in the BOSH Agent: #243
- Some adjustment to the BOSH Director logic: cloudfoundry/bosh#2318

#### This PR relates to
- cloudfoundry/bosh-aws-cpi-release#109, that can be tested and merged when the current work is done.

#### Context

- The director property `enable_cpi_resize_disk` enables this feature:
 https://bosh.io/jobs/director?source=github.com/cloudfoundry/bosh&version=271.9.0#p%3ddirector.enable_cpi_resize_disk

- `resize disk` PR for the AWS CPI, which includes discussions about disk resizing: [Add a new API method resize_disk](https://github.com/cloudfoundry/bosh-aws-cpi-release/pull/109)

- bosh-agent issue [Resizing disk with enable_cpi_resize_disk option gets error...](https://github.com/cloudfoundry/bosh-agent/issues/221)

- GitHub issue [disk resize miss linux fs resize](https://github.com/cloudfoundry/bosh-openstack-cpi-release/issues/161)

- Closed bosh-agent PR: https://github.com/cloudfoundry/bosh-agent/pull/233
 bosh-agent WIP commit:
 platform/disk/ext4_file_system_extender.go

- Bosh tracker story about investigation when we realized openstack resize was broken: https://www.pivotaltracker.com/n/projects/1456570/stories/161478343

- Original exploration doc around resizing: https://docs.google.com/document/d/1Eoa1UPUtU6m4jPjJGqsE4jwajmirELql6gL4zSxeN3k/edit?usp=sharing

### CODE REVIEW

**Comment by bgandon** [comment]
I've reviewed and improved test structure and code coverage. This now looks good to me.
I've also bumped the `bosh-utils` dependency in order to include the fix from cloudfoundry/bosh-utils#80.

File: platform/disk/parted_partitioner.go [hunk 1]
```diff
@@ -370,3 +370,86 @@ func (p partedPartitioner) createMapperPartition(devicePath string) error {
...
+			return true, nil
+		}
+
+		remainingDiskSpace = remainingDiskSpace - partitionToMatch.SizeInBytes
+	}
+
+	return false, nil
+}
+
+func (p partedPartitioner) ResizePartitions(devicePath string, partitionsToMatch []Partition) (err error) {
```
Comment (rkoster) [inline platform/disk/parted_partitioner.go:407]: This function does not need `partitionsToMatch` it is only used to check the SizeInBytes which is never set by the caller anyway.

**Review by rkoster — CHANGES_REQUESTED** [review]
Looks good overall, just 2 functions that could be simplified.

File: platform/disk/parted_partitioner.go [hunk 2]
```diff
@@ -370,3 +370,86 @@ func (p partedPartitioner) createMapperPartition(devicePath string) error {
 	detectPartitionRetryStrategy := NewPartitionStrategy(detectPartitionRetryable, p.timeService, p.logger)
 	return detectPartitionRetryStrategy.Try()
 }
+
+func (p partedPartitioner) PartitionsNeedResize(devicePath string, partitionsToMatch []Partition) (needsResize bool, err error) {
```
[the remark this answers is not in the document]
Comment (bgandon) [reply platform/disk/parted_partitioner.go:374]: That's right indeed.

File: platform/disk/parted_partitioner.go (hunk 1 above)
Comment (bgandon) [reply platform/disk/parted_partitioner.go:407]: All right, I'll rework the signature of this method and simplify the algorithm.

**Comment by bgandon** [comment]
Ok, there we are. I've simplified the method signatures, the algorithms, and the unit tests.

**Review by rkoster — APPROVED** [review]
LGTM @bgandon Thanks!
cc @camilalonart

**Pull request merged**